loader/src/main/java/net/neoforged/fml/ModLoader.java

```diff
@@ -190,22 +191,58 @@ public static void waitForTask(String name, Runnable periodicTask, CompletableFu
         }
     }
 
+    /**
+     * Exception that is fired when a mod loading future cannot be executed because a dependent future failed.
+     * It is only used for control flow and easy filtering out, but never logged or propagated further.
+     */
+    private static class DependentFutureFailedException extends RuntimeException {}
+
     /**
      * Dispatches a task across all mod containers in parallel, with progress displayed on the loading screen.
      */
     public static void dispatchParallelTask(String name, Executor parallelExecutor, Runnable periodicTask, Consumer<ModContainer> task) {
         var progress = StartupNotificationManager.addProgressBar(name, modList.size());
         try {
             periodicTask.run();
+            Map<IModInfo, CompletableFuture<Void>> modFutures = new IdentityHashMap<>();
             var futureList = modList.getSortedMods().stream()
                     .map(modContainer -> {
-                        return CompletableFuture.runAsync(() -> {
-                            ModLoadingContext.get().setActiveContainer(modContainer);
-                            task.accept(modContainer);
-                        }, parallelExecutor).whenComplete((result, exception) -> {
-                            progress.increment();
-                            ModLoadingContext.get().setActiveContainer(null);
-                        });
+                        // Build future for all dependencies first
+                        var deps = LoadingModList.get().getDependencies(modContainer.getModInfo());
+                        @SuppressWarnings("unchecked")
+                        CompletableFuture<Void>[] depFutures = new CompletableFuture[deps.size()];
+                        for (int i = 0; i < deps.size(); ++i) {
+                            depFutures[i] = modFutures.get(deps.get(i));
+                            if (depFutures[i] == null) {
+                                throw new IllegalStateException("Dependency future for mod %s which is a dependency of %s not found!".formatted(
+                                        deps.get(i).getModId(), modContainer.getModId()));
+                            }
+                        }
+
+                        var combinedFuture = depFutures.length == 0 ? CompletableFuture.completedFuture(null) : CompletableFuture.allOf(depFutures);
```

Comment:
You can just use `CompletableFuture.allOf` unconditionally; it [handles](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html#allOf-java.util.concurrent.CompletableFuture...-) zero-length arrays.

Reply:
Done :)